marshmallow_utils/schemas/identifier.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 CERN.
+#
+# Marshmallow-Utils is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Identifiers Schema's."""
+
+import idutils
+from marshmallow import Schema, ValidationError, pre_load, validates_schema
+
+from ..fields import SanitizedUnicode
+
+
+class IdentifierSchema(Schema):
+    """Identifier with automatic scheme detection."""
+
+    identifier = SanitizedUnicode()
+    scheme = SanitizedUnicode()
+
+    def __init__(self, allowed_schemes=None, allow_all=False,
+                 required=True, **kwargs):
+        """Constructor.
+
+        `allowed_schemas` is incompatible with `allow_all`.
+        If `allow_all` is set to `True` the `allowed_schemes` will ignored.
+
+        The `required` param applies to the `identifier` value.
+        """
+        self.allow_all = allow_all
+        self.allowed_schemes = None if allow_all else allowed_schemes
+        self.required = required
+        super().__init__(**kwargs)
+
+    @pre_load(pass_many=False)
+    def detect_schema(self, data, **kwargs):
+        """Detects the schema of the identifier."""
+        identifier = data.get("identifier")
+        scheme = data.get("scheme")
+
+        # Identifier is a required field, but pre-load happens before
+        # It will fail on first step validation
+        if not identifier:
```

Comment:
Minor: I think this method is trying to do two things a bit mixed: 1) detect the scheme if it's not provided and 2) normalise the identifier.

I think (1) is suitable for ``pre_load``, but I would expect (2) to happen on ``post_load()`` - i.e. given that we have a valid identifier, then we normalise the identifier.

Reply:
Taking above comment into account, I would bail out here if 1) identifier is not provided or 2) scheme and identifier is provided.

Below code should then only focus on detection of the identifier scheme.

Reply:
Thanks! I agree, I refactored to do detection on pre load, validation, then normalization on post load. However, I do not understand the second comment about "bailing"?